jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAggregator.java

```diff
@@ -0,0 +1,138 @@
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>Aggregates data into a single ByteBuffer of a specified maximum size.</p>
+ * <p>The buffer automatically grows as data is written to it, up until it reaches the specified maximum size.
+ * Once the buffer is full, the aggregator will not aggregate any more bytes until its buffer is taken out,
+ * after which a new aggregate/take buffer cycle can start.</p>
+ * <p>The buffers are taken from the supplied {@link ByteBufferPool} or freshly allocated if one is not supplied.</p>
+ */
+public class ByteBufferAggregator
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ByteBufferAggregator.class);
+
+    private final ByteBufferPool _bufferPool;
+    private final boolean _direct;
+    private final int _maxSize;
+    private RetainableByteBuffer _retainableByteBuffer;
+    private int _aggregatedSize;
+    private int _currentSize;
+
+    /**
+     * Creates a ByteBuffer aggregator.
+     * @param bufferPool The {@link ByteBufferPool} from which to acquire the buffers
+     * @param direct whether to get direct buffers
+     * @param startSize the starting size of the buffer
+     * @param maxSize the maximum size of the buffer which must be greater than {@code startSize}
+     */
+    public ByteBufferAggregator(ByteBufferPool bufferPool, boolean direct, int startSize, int maxSize)
+    {
+        if (maxSize <= 0)
+            throw new IllegalArgumentException("maxSize must be > 0, was: " + maxSize);
+        if (startSize <= 0)
+            throw new IllegalArgumentException("startSize must be > 0, was: " + startSize);
+        if (startSize > maxSize)
+            throw new IllegalArgumentException("maxSize (" + maxSize + ") must be >= startSize (" + startSize + ")");
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxSize = maxSize;
+        _currentSize = startSize;
+    }
+
+    /**
+     * Copies the given ByteBuffer into this aggregator. This will aggregate bytes up to the
+     * specified maximum size, at which time this method returns false and
+     * {@link #takeRetainableByteBuffer()} must be called for this method to accept aggregating again.
+     * @param buffer the buffer to copy into this aggregator; its position is updated according to
+     * the number of aggregated bytes
+     * @return true if the aggregator's buffer is full and should be taken, false otherwise
+     */
+    public boolean aggregate(ByteBuffer buffer)
```

Comment:
I would have preferred a clearer semantic of "if I can't aggregate it all, I'll leave it untouched", rather than aggregating few bytes and leaving the buffer half-consumed.

Reply:
I'm sure there are heuristics that could be applied here about when is it better to:
 a) aggregate a partial buffer; write aggregate; aggregate remainder
 b) write aggregate; aggregate buffer
 c) write aggregate; write remainder

Always doing one or the other will always have examples when the other is better.  But I think a)  is probably a reasonable start.      Or spend an hour to come up with some heuristic rough choice between the 3 options.

Reply:
I'm going to keep the contract as it currently is, stating that the given buffer might be partially consumed so that we have more flexibility in the future to change the implementation.